File: backend/get-metar.js
// Patrick Clarke
// get-metar.js
// 3/1/'26

// To run: node get-metar.js

/*
// Actual list of 165 weather stations.
const californiaStationIDs = [
  "KNGZ", // Alameda NAS, CA
  "KAAT", // Alturas, CA
  "KAPV", // Apple Valley, CA
  "KACV", // Arcata/Eureka, CA
  "KAUN", // Auburn Muni, CA
  "KAVX", // Avalon, CA
  "KBFL", // Bakersfield, CA
  "KBAB", // Beale AFB/Marysv, CA
  "KBUO", // Beaumont, CA
  "KBYS", // Bicycle Lake, CA
  "KL35", // Big Bear City, CA
  "KBIH", // Bishop, CA
  "KBLH", // Blythe, CA
  "KL08", // Borrego Springs, CA
  "KO57", // Bridgeport, CA
  "KBAN", // Bridgeport(Mcmwt, CA
  "KBUR", // Burbank, CA
  "KBNY", // Burney, CA
  "KC83", // Byron, CA
  "KCXL", // Calexico Intl, CA
  "KCMA", // Camarillo, CA
  "KCZZ", // Campo, CA
  "KNFG", // Camp Pendleton, CA
  "KCRQ", // Carlsbad, CA
  "KO59", // Cedarville, CA
  "KCIC", // Chico Municipal, CA
  "KNID", // China Lake (NAF), CA
  "KCNO", // Chino, CA
  "KO22", // Columbia, CA
  "KCCR", // Concord, CA
  "KAJO", // Corona Muni, CA
  "KSNA", // Costa Mesa, CA
  "KCEC", // Crescent City, CA
  "KDAG", // Daggett, CA
  "KDWA", // Davis Yolo Cnty, CA
  "KEDU", // Davis, CA
  "KDLO", // Delano, CA
  "KEDW", // Edwards AFB, CA
  "K9L2", // Edwards N-Aux, CA
  "KNJK", // El Centro NAF, CA
  "KEMT", // El Monte, CA
  "KBLU", // Emigrant Gap, CA
  "KEKA", // Eureka/Murray, CA
  "KL18", // Fallbrook Airpar, CA
  "KFOT", // Fortuna, CA
  "KFCH", // Fresno Chandler, CA
  "KFAT", // Fresno, CA
  "KFUL", // Fullerton, CA
  "KGOO", // Grass Valley, CA
  "KHAF", // Half Moon Bay, CA
  "KHJO", // Hanford, CA
  "KO18", // Hanford, CA
  "KHHR", // Hawthorne, CA
  "KHWD", // Hayward, CA
  "KHES", // Healdsburg, CA
  "KHMT", // Hemet, CA
  "KCVH", // Hollister Muni, CA
  "KHGT", // Hunter Ligget, CA
  "KNRS", // Imperial Beach, CA
  "KIPL", // Imperial, CA
  "KIYK", // Inyokern, CA
  "KJAQ", // Jackson, CA
  "KWJF", // Lancaster/Fox, CA
  "KPOC", // La Verne/Bracket, CA
  "KWHP", // La / Whiteman, CA
  "KNLC", // Lemoore NAS/Reev, CA
  "KLHM", // Lincoln, CA
  "KLLR", // Little River, CA
  "KLVK", // Livermore, CA
  "KLPC", // Lompoc, CA
  "KLGB", // Long Beach, CA
  "KSLI", // Los Alamitos AAF, CA
  "KCQT", // Los Angeles, CA
  "KLAX", // Los Angeles, CA
  "KMAE", // Madera, CA
  "KMMH", // Mammoth/June Lak, CA
  "KMYV", // Marysville, CA
  "KMHR", // Mather Field, CA
  "KMCC", // Mcclellan AFB, CA
  "KMER", // Merced/Castle AF, CA
  "KMCE", // Merced, CA
  "KNKX", // Miramar NAS/San, CA
  "KMOD", // Modesto, CA
  "KNUQ", // Moffett NAS/Mtn, CA
  "KMHV", // Mojave, CA
  "KSIY", // Montague/Siskiyo, CA
  "KMRY", // Monterey, CA
  "KMHS", // Mount Shasta, CA
  "KMWS", // Mount Wilson, CA
  "KF70", // Murrieta/Temecul, CA
  "KAPC", // Napa, CA
  "KEED", // Needles, CA
  "K3A6", // Newhall, CA
  "KNZY", // North Island NAS, CA
  "KDVO", // Novato/Gnoss Fld, CA
  "KOAK", // Oakland, CA
  "KL52", // Oceano Cnty, CA
  "KOKB", // Oceanside, CA
  "KNXF", // Oceanside/Red-B, CA
  "KONT", // Ontario, CA
  "KOVE", // Oroville, CA
  "KOXR", // Oxnard, CA
  "KGXA", // Palmdale, CA
  "KPMD", // Palmdale, CA
  "KPSP", // Palm Springs, CA
  "KPAO", // Palo Alto, CA
  "KPRB", // Paso Robles, CA
  "KO69", // Petaluma, CA
  "KPVF", // Placerville, CA
  "KNTD", // Point Mugu NAS, CA
  "KPTV", // Porterville, CA
  "K87Q", // Pt. Piedras Blan, CA
  "KRNM", // Ramona, CA
  "KRBL", // Red Bluff, CA
  "KRDD", // Redding, CA
  "KREI", // Redlands, CA
  "KO32", // Reedley, CA
  "KO88", // Rio Vista, CA
  "KRAL", // Riverside, CA
  "KRIV", // Riverside/March, CA
  "KSAC", // Sacramento, CA
  "KSMF", // Sacramento/Metro, CA
  "KSNS", // Salinas, CA
  "KCPU", // San Andreas, CA
  "KSBD", // San Bernardino, CA
  "KSQL", // San Carlos Airpo, CA
  "KNUC", // San Clemente Is., CA
  "KSDB", // Sandberg, CA
  "KSDM", // San Diego/Brown, CA
  "KSAN", // San Diego, CA
  "KMYF", // San Diego/Mntgmy, CA
  "KSEE", // San Diego/Santee, CA
  "KSFO", // San Francisco, CA
  "KSJC", // San Jose, CA
  "KRHV", // San Jose/Reid, CA
  "KSBP", // San Luis Obispo, CA
  "KE16", // San Martin, CA
  "KNSI", // San Nicolas Isla, CA
  "KSBA", // Santa Barbara, CA
  "KSMX", // Santa Maria, CA
  "KSMO", // Santa Monica, CA
  "KSTS", // Santa Rosa, CA
  "KIZA", // Santa Ynez, CA
  "KO87", // Shelter Cove, CA
  "KTVL", // South Lake Tahoe, CA
  "KSCK", // Stockton, CA
  "KSVE", // Susanville Muni, CA
  "KTSP", // Tehachapi, CA
  "KTRM", // Thermal/Palm Spg, CA
  "KTOA", // Torrance Municip, CA
  "KTCY", // Tracy, CA
  "KSUU", // Travis AFB/Fairf, CA
  "KO86", // Trinity Center, CA
  "KTRK", // Truckee Tahoe, CA
  "KNXP", // Twentynine Palms, CA
  "KUKI", // Ukiah, CA
  "KCCB", // Upland, CA
  "KVCB", // Vacaville, CA
  "KVBG", // Vandenberg AFB, CA
  "KXVW", // Vandenberg Range, CA
  "KVNY", // Van Nuys, CA
  "KVCV", // Victorville, CA
  "KVIS", // Visalia Muni, CA
  "KWVI", // Watsonville, CA
  "KO54"  // Weaverville, CA
];
*/

// Small list of 10 to test.
const californiaStationIDs = [
  "KNGZ", // Alameda NAS, CA
  "KAAT", // Alturas, CA
  "KAPV", // Apple Valley, CA
  "KACV", // Arcata/Eureka, CA
  "KAUN", // Auburn Muni, CA
  "KAVX", // Avalon, CA
  "KBFL", // Bakersfield, CA
  "KBAB", // Beale AFB/Marysv, CA
  "KBUO", // Beaumont, CA
  "KBYS", // Bicycle Lake, CA
  "KL35", // Big Bear City, CA
];


// Printing an individual array entry.
// console.log(californiaStationIDs[0]);

const fs = require('fs');

async function fetchAllWeather() {
  const allWeatherData = {};
  
  // NWS REQUIRES a User-Agent header. Change this to your own info.
  const headers = {
    'User-Agent': 'MyWeatherApp/1.0 (your-email@example.com)',
    'Accept': 'application/geo+json'
  };

  console.log(`Starting fetch for ${californiaStationIDs.length} stations...`);

  // 2. Loop through each ID sequentially
  for (const id of californiaStationIDs) {
    try {
      console.log(`Fetching data for ${id}...`);
      const response = await fetch(`https://api.weather.gov/stations/${id}/observations/latest`, { headers });
      
      if (!response.ok) {
        throw new Error(`HTTP Error: ${response.status}`);
      }

      const data = await response.json();
      allWeatherData[id] = data; // Store the data under the station's ID

      // 3. Pause for 500 milliseconds to respect NWS rate limits
      await new Promise(resolve => setTimeout(resolve, 500));

    } catch (error) {
      console.error(`Failed to fetch ${id}: ${error.message}`);
      allWeatherData[id] = { error: error.message }; // Log the error in the JSON so it doesn't break the whole process
    }
  }

  // 4. Save the compiled object to a JSON file
  fs.writeFileSync('california_weather.json', JSON.stringify(allWeatherData, null, 2));
  console.log('✅ Success! Data saved to california_weather.json');
}

// Run the function
fetchAllWeather();

File: backend/get-metar.py
# # Patrick Clarke
# # get_metar.py
# # 3/22/'26

# # METAR stands for METeorological Aerodrome Report

# # To run: python get_metar.py

# # uvicorn is the server
# # fastapi is the framework.
# # (Note: This script runs standalone, but can easily be imported into a FastAPI route later)

# import json
# import time
# import requests

# """
# There are 22 weather station IDs returning a 404 Error as of 3/4/'26:

# california_404_station_ids = [
#     "KNGZ", # Alameda NAS, CA
#     "KBUO", # Beaumont, CA
#     "KO57", # Bridgeport, CA
#     "KBNY", # Burney, CA
#     "KCXL", # Calexico Intl, CA
#     "KCZZ", # Campo, CA
#     "KO59", # Cedarville, CA
#     "K9L2", # Edwards N-Aux, CA
#     "KEKA", # Eureka/Murray, CA
#     "KO18", # Hanford, CA *There are two Hanfords so this one is likely deprecated.
#     "KHES", # Healdsburg, CA
#     "KHGT", # Hunter Ligget, CA
#     "KCQT", # Los Angeles, CA
#     "K3A6", # Newhall, CA
#     "KL52", # Oceano Cnty, CA
#     "KNXF", # Oceanside/Red-B, CA
#     "KGXA", # Palmdale, CA *There are also two Palmdales so this one is also likely deprecated.
#     "K87Q", # Pt. Piedras Blan, CA
#     "KO88", # Rio Vista, CA
#     "KO87", # Shelter Cove, CA
#     "KTRM", # Thermal/Palm Spg, CA
#     "KXVW", # Vandenberg Range, CA
# ]
# """

# # Actual list of 165 weather stations.
# california_station_ids = [
#     "KNGZ", "KAAT", "KAPV", "KACV", "KAUN", "KAVX", "KBFL", "KBAB", "KBUO", "KBYS", 
#     "KL35", "KBIH", "KBLH", "KL08", "KO57", "KBAN", "KBUR", "KBNY", "KC83", "KCXL", 
#     "KCMA", "KCZZ", "KNFG", "KCRQ", "KO59", "KCIC", "KNID", "KCNO", "KO22", "KCCR", 
#     "KAJO", "KSNA", "KCEC", "KDAG", "KDWA", "KEDU", "KDLO", "KEDW", "K9L2", "KNJK", 
#     "KEMT", "KBLU", "KEKA", "KL18", "KFOT", "KFCH", "KFAT", "KFUL", "KGOO", "KHAF", 
#     "KHJO", "KO18", "KHHR", "KHWD", "KHES", "KHMT", "KCVH", "KHGT", "KNRS", "KIPL", 
#     "KIYK", "KJAQ", "KWJF", "KPOC", "KWHP", "KNLC", "KLHM", "KLLR", "KLVK", "KLPC", 
#     "KLGB", "KSLI", "KCQT", "KLAX", "KMAE", "KMMH", "KMYV", "KMHR", "KMCC", "KMER", 
#     "KMCE", "KNKX", "KMOD", "KNUQ", "KMHV", "KSIY", "KMRY", "KMHS", "KMWS", "KF70", 
#     "KAPC", "KEED", "K3A6", "KNZY", "KDVO", "KOAK", "KL52", "KOKB", "KNXF", "KONT", 
#     "KOVE", "KOXR", "KGXA", "KPMD", "KPSP", "KPAO", "KPRB", "KO69", "KPVF", "KNTD", 
#     "KPTV", "K87Q", "KRNM", "KRBL", "KRDD", "KREI", "KO32", "KO88", "KRAL", "KRIV", 
#     "KSAC", "KSMF", "KSNS", "KCPU", "KSBD", "KSQL", "KNUC", "KSDB", "KSDM", "KSAN", 
#     "KMYF", "KSEE", "KSFO", "KSJC", "KRHV", "KSBP", "KE16", "KNSI", "KSBA", "KSMX", 
#     "KSMO", "KSTS", "KIZA", "KO87", "KTVL", "KSCK", "KSVE", "KTSP", "KTRM", "KTOA", 
#     "KTCY", "KSUU", "KO86", "KTRK", "KNXP", "KUKI", "KCCB", "KVCB", "KVBG", "KXVW", 
#     "KVNY", "KVCV", "KVIS", "KWVI", "KO54"
# ]

# """
# # Small list of 10 to test.
# california_station_ids = [
#     "KNGZ", # Alameda NAS, CA
#     "KAAT", # Alturas, CA
#     "KAPV", # Apple Valley, CA
#     "KACV", # Arcata/Eureka, CA
#     "KAUN", # Auburn Muni, CA
#     "KAVX", # Avalon, CA
#     "KBFL", # Bakersfield, CA
#     "KBAB", # Beale AFB/Marysv, CA
#     "KBUO", # Beaumont, CA
#     "KBYS", # Bicycle Lake, CA
#     "KL35", # Big Bear City, CA
# ]
# """

# # Printing an individual array entry.
# # print(california_station_ids[0])

# def fetch_all_weather():
#     all_weather_data = {}
    
#     # NWS REQUIRES a User-Agent header. Change this to your own info.
#     headers = {
#         'User-Agent': 'MyWeatherApp/1.0 (your-email@example.com)',
#         'Accept': 'application/geo+json'
#     }

#     print(f"Starting fetch for {len(california_station_ids)} stations...")

#     # 2. Loop through each ID sequentially
#     for station_id in california_station_ids:
#         try:
#             print(f"Fetching data for {station_id}...")

#             # Fetch the recent observations instead of just the latest
#             url = f"https://api.weather.gov/stations/{station_id}/observations"
#             response = requests.get(url, headers=headers)
            
#             # Check if the HTTP request was successful before trying to parse JSON
#             response.raise_for_status() 
            
#             data = response.json()

#             # Find the first (newest) observation where the temperature is NOT null
#             # Using next() with a generator expression replaces JS's Array.prototype.find()
#             best_observation = next(
#                 (obs for obs in data.get('features', []) 
#                  if obs.get('properties', {}).get('temperature', {}).get('value') is not None), 
#                 None
#             )

#             # Store that good observation!
#             all_weather_data[station_id] = best_observation

#             # 3. Pause for 500 milliseconds (0.5 seconds) to respect NWS rate limits
#             time.sleep(0.5)

#         except requests.exceptions.RequestException as error:
#             print(f"Failed to fetch {station_id}: {error}")
#             all_weather_data[station_id] = {"error": str(error)}
#         except Exception as error:
#             print(f"An unexpected error occurred for {station_id}: {error}")
#             all_weather_data[station_id] = {"error": str(error)}

#     # 4. Save the compiled dictionary to a JSON file
#     with open('california_weather.json', 'w', encoding='utf-8') as f:
#         json.dump(all_weather_data, f, indent=2)
        
#     print('✅ Success! Data saved to california_weather.json')

# # Run the function
# if __name__ == "__main__":
#     fetch_all_weather()



# Patrick Clarke
# get_metar.py
# 3/22/'26

import json
import asyncio
import aiohttp
import time

"""
This is an improved version of the original file that gets weather data in 40s - 1m
compared to three minutes. 
"""

california_station_ids = [
    "KNGZ", "KAAT", "KAPV", "KACV", "KAUN", "KAVX", "KBFL", "KBAB", "KBUO", "KBYS", 
    "KL35", "KBIH", "KBLH", "KL08", "KO57", "KBAN", "KBUR", "KBNY", "KC83", "KCXL", 
    "KCMA", "KCZZ", "KNFG", "KCRQ", "KO59", "KCIC", "KNID", "KCNO", "KO22", "KCCR", 
    "KAJO", "KSNA", "KCEC", "KDAG", "KDWA", "KEDU", "KDLO", "KEDW", "K9L2", "KNJK", 
    "KEMT", "KBLU", "KEKA", "KL18", "KFOT", "KFCH", "KFAT", "KFUL", "KGOO", "KHAF", 
    "KHJO", "KO18", "KHHR", "KHWD", "KHES", "KHMT", "KCVH", "KHGT", "KNRS", "KIPL", 
    "KIYK", "KJAQ", "KWJF", "KPOC", "KWHP", "KNLC", "KLHM", "KLLR", "KLVK", "KLPC", 
    "KLGB", "KSLI", "KCQT", "KLAX", "KMAE", "KMMH", "KMYV", "KMHR", "KMCC", "KMER", 
    "KMCE", "KNKX", "KMOD", "KNUQ", "KMHV", "KSIY", "KMRY", "KMHS", "KMWS", "KF70", 
    "KAPC", "KEED", "K3A6", "KNZY", "KDVO", "KOAK", "KL52", "KOKB", "KNXF", "KONT", 
    "KOVE", "KOXR", "KGXA", "KPMD", "KPSP", "KPAO", "KPRB", "KO69", "KPVF", "KNTD", 
    "KPTV", "K87Q", "KRNM", "KRBL", "KRDD", "KREI", "KO32", "KO88", "KRAL", "KRIV", 
    "KSAC", "KSMF", "KSNS", "KCPU", "KSBD", "KSQL", "KNUC", "KSDB", "KSDM", "KSAN", 
    "KMYF", "KSEE", "KSFO", "KSJC", "KRHV", "KSBP", "KE16", "KNSI", "KSBA", "KSMX", 
    "KSMO", "KSTS", "KIZA", "KO87", "KTVL", "KSCK", "KSVE", "KTSP", "KTRM", "KTOA", 
    "KTCY", "KSUU", "KO86", "KTRK", "KNXP", "KUKI", "KCCB", "KVCB", "KVBG", "KXVW", 
    "KVNY", "KVCV", "KVIS", "KWVI", "KO54"
]

# NWS REQUIRES a User-Agent header. Change this to your own info.
HEADERS = {
    'User-Agent': 'MyWeatherApp/1.0 (your-email@example.com)',
    'Accept': 'application/geo+json'
}

async def fetch_station(session, station_id, semaphore):
    """Fetches data for a single station, gated by the semaphore."""
    # The semaphore ensures only X number of these run at the exact same time
    async with semaphore:
        url = f"https://api.weather.gov/stations/{station_id}/observations"
        try:
            # We add a tiny artificial delay to ensure we don't trigger rate limits
            await asyncio.sleep(0.2) 
            
            #async with session.get(url, headers=HEADERS) as response:
            async with session.get(url, headers=HEADERS, ssl=False) as response:
                response.raise_for_status() # Throw error if not 200 OK
                data = await response.json()

                # Find the first (newest) observation where the temperature is NOT null
                best_observation = next(
                    (obs for obs in data.get('features', []) 
                     if obs.get('properties', {}).get('temperature', {}).get('value') is not None), 
                    None
                )
                
                print(f"✅ Fetched {station_id}")
                return station_id, best_observation

        except aiohttp.ClientResponseError as error:
            print(f"❌ HTTP Error for {station_id}: {error.status}")
            return station_id, {"error": str(error)}
        except Exception as error:
            print(f"❌ Unexpected Error for {station_id}: {error}")
            return station_id, {"error": str(error)}

async def main():
    start_time = time.time()
    all_weather_data = {}
    
    # Set the Semaphore limit. 5 is a very safe number for the NWS API.
    # It means 5 concurrent requests at a time.
    semaphore = asyncio.Semaphore(5)
    
    print(f"Starting async fetch for {len(california_station_ids)} stations...")

    # Create a single persistent session for all requests (much faster than opening new connections)
    async with aiohttp.ClientSession() as session:
        # Create a list of "tasks" (promises) to run
        tasks = [fetch_station(session, station_id, semaphore) for station_id in california_station_ids]
        
        # Run them all together and wait for them to finish
        results = await asyncio.gather(*tasks)
        
        # Rebuild the dictionary from the results
        for station_id, data in results:
            all_weather_data[station_id] = data

    # Save to JSON
    with open('california_weather.json', 'w', encoding='utf-8') as f:
        json.dump(all_weather_data, f, indent=2)
        
    end_time = time.time()
    print(f"\n Success! Data saved to california_weather.json")
    print(f"Total execution time: {round(end_time - start_time, 2)} seconds")

# # Run the async event loop (Windows user)
# if __name__ == "__main__":
#     # Windows sometimes throws a harmless but annoying error at the end of asyncio scripts.
#     # This specific setup prevents it.
#     asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
#     asyncio.run(main())

# Run the async event loop
if __name__ == "__main__":
    asyncio.run(main())